src/sql/src/query_model/rewrite_engine.rs

```diff
@@ -0,0 +1,969 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use crate::query_model::{
+    BoxId, BoxScalarExpr, BoxType, ColumnReference, DistinctOperation, Model, QuantifierId,
+    QuantifierSet, QuantifierType, Select,
+};
+use itertools::Itertools;
+use std::collections::{BTreeMap, BTreeSet, HashMap, HashSet};
+
+/// Rule type
+#[allow(dead_code)]
+#[derive(Debug)]
+pub enum RuleType {
+    TopBoxOnly,
+    PreOrder,
+    PostOrder,
+}
+
+/// Trait for rewrite rules
+pub trait Rule {
+    fn name(&self) -> &'static str;
+
+    fn rule_type(&self) -> RuleType;
+
+    /// Whether the action should be fired for the given box.
+    /// This method is not allowed to modify the model in any way.
+    fn condition(&mut self, model: &Model, box_id: BoxId) -> bool;
+
+    /// Invoked immediately after `condition` if it returned true.
+    fn action(&mut self, model: &mut Model, box_id: BoxId);
+}
+
+/// Entry-point of the normalization stage.
+pub fn rewrite_model(model: &mut Model) {
+    let mut fixup_rules: Vec<Box<dyn Rule>> = vec![Box::new(Windowing::new())];
+    apply_rules_to_model(model, &mut fixup_rules);
+    model.garbage_collect();
+
+    let mut rules: Vec<Box<dyn Rule>> = vec![
```

Comment:
Eventually it won't matter, since we will keep traversing the graph until a full traversal is performed without a single rule being fired.

At the moment, since we are only traversing the graph once, we may not reach that stable point in a single pass, so a different order of the rules may lead to a different output.

The rewrite engine needs more work and planning, but what I wanted to show in this branch is how the rules could look like, specially the split between the `condition` and the `action` method.